pkg/controller/resize_status.go

```diff
@@ -0,0 +1,111 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+// markControllerResizeInProgress will mark PVC for controller resize, this function is newer version that uses
+// resizeStatus and sets allocatedResources.
+func (ctrl *resizeController) markControllerResizeInProgress(
+	pvc *v1.PersistentVolumeClaim, newSize resource.Quantity) (*v1.PersistentVolumeClaim, error) {
+
+	progressCondition := v1.PersistentVolumeClaimCondition{
+		Type:               v1.PersistentVolumeClaimResizing,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+	}
+	controllerExpansionInProgress := v1.PersistentVolumeClaimControllerExpansionInProgress
+	conditions := []v1.PersistentVolumeClaimCondition{progressCondition}
+
+	newPVC := pvc.DeepCopy()
+	newPVC.Status.Conditions = util.MergeResizeConditionsOfPVC(newPVC.Status.Conditions, conditions)
+	newPVC.Status.ResizeStatus = &controllerExpansionInProgress
+	newPVC.Status.AllocatedResources = v1.ResourceList{v1.ResourceStorage: newSize}
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, true /* addResourceVersionCheck */)
+	if err != nil {
+		return nil, err
+	}
+	return updatedPVC, nil
+}
+
+// markForPendingNodeExpansion is new set of functions designed around feature RecoverVolumeExpansionFailure
+// which correctly sets pvc.Status.ResizeStatus
+func (ctrl *resizeController) markForPendingNodeExpansion(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	pvcCondition := v1.PersistentVolumeClaimCondition{
+		Type:               v1.PersistentVolumeClaimFileSystemResizePending,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+		Message:            "Waiting for user to (re-)start a pod to finish file system resize of volume on node.",
+	}
+
+	nodeExpansionPendingVar := v1.PersistentVolumeClaimNodeExpansionPending
+	newPVC := pvc.DeepCopy()
+	newPVC.Status.Conditions = util.MergeResizeConditionsOfPVC(newPVC.Status.Conditions,
+		[]v1.PersistentVolumeClaimCondition{pvcCondition})
+	newPVC.Status.ResizeStatus = &nodeExpansionPendingVar
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, true /* addResourceVersionCheck */)
+
+	if err != nil {
+		return updatedPVC, fmt.Errorf("Mark PVC %q as node expansion required failed: %v", util.PVCKey(pvc), err)
+	}
+
+	klog.V(4).Infof("Mark PVC %q as file system resize required", util.PVCKey(pvc))
+	ctrl.eventRecorder.Eventf(pvc, v1.EventTypeNormal,
+		util.FileSystemResizeRequired, "Require file system resize of volume on node")
+
+	return updatedPVC, nil
+}
+
+func (ctrl *resizeController) markControllerExpansionFailed(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	expansionFailedOnController := v1.PersistentVolumeClaimControllerExpansionFailed
+	newPVC := pvc.DeepCopy()
+	newPVC.Status.ResizeStatus = &expansionFailedOnController
+
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, false /* addResourceVersionCheck */)
```

Comment:
the reason no need to check resource version here?

Reply:
The general idea I had was if expansion failed we record the error even if we get a potential conflict. Since we are only patching  resize status we should be okay.

Reply:
what is the result of conflict? It might overwrite some fields which are updated by other places?
how about we also check addResourceVersionCheck here? Any potential issue?

Reply:
The conflict when `addResourceVersionCheck` is present has nothing to do with content of the patch, but is about whether patch we are applying is with latest version of object or not. i.e - if two updates were performed on PVC concurrently and they updated different fields, then if `addResourceVersionCheck` is present with second patch - you will still get error (even if second update, changes entirely different field of PVC).  

> how about we also check addResourceVersionCheck here? Any potential issue?

We can, but if somehow our version of PVC was older then, the entire resizing operation has to be restarted before `ExpansionFailedOnController` can be set for `resizeStatus`. As such it does not affect the design but skipping `addResourceVersionCheck` is an optimization - so as we set this field if expansion fails on controller even if PVC was updated elsewhere.

Reply:
Added as a comment in the code.